src/pod-security-policy/seccomp/src.rego

```diff
@@ -2,40 +2,88 @@ package k8spspseccomp
-    msg := sprintf("Seccomp profile is not allowed, pod: %v, container: %v, Allowed profiles: %v", [metadata.name, container.name, input.parameters.allowedProfiles])
-}
+container_annotation_key_prefix = "container.seccomp.security.alpha.kubernetes.io/"
+
+pod_annotation_key = "seccomp.security.alpha.kubernetes.io/pod"
 
-input_wildcard_allowed(metadata) {
-    input.parameters.allowedProfiles[_] == "*"
+violation[{"msg": msg}] {
+	not input_wildcard_allowed
+	container := input_containers[name]
+	not is_exempt(container)
+	result := get_profile(container)
+	not allowed_profile(result.profile)
+	msg := sprintf("Seccomp profile '%v' is not allowed for container '%v'. Found at: %v. Allowed profiles: %v", [result.profile, name, result.location, input.parameters.allowedProfiles])
 }
 
-input_container_allowed(metadata, container) {
-    not get_container_profile(metadata, container)
-    metadata.annotations["seccomp.security.alpha.kubernetes.io/pod"] == input.parameters.allowedProfiles[_]
+input_wildcard_allowed {
+	input.parameters.allowedProfiles[_] == "*"
 }
 
-input_container_allowed(metadata, container) {
-	profile := get_container_profile(metadata, container)
+allowed_profile(profile) {
 	profile == input.parameters.allowedProfiles[_]
 }
 
-get_container_profile(metadata, container) = profile {
-	value := metadata.annotations[key]
-    startswith(key, "container.seccomp.security.alpha.kubernetes.io/")
-    [prefix, name] := split(key, "/")
-    name == container.name
-    profile = value
+# Container profile as defined in pod annotation
+get_profile(container) = {"profile": profile, "location": location} {
+	not has_securitycontext_container(container)
+	not has_annotation(get_container_annotation_key(container.name))
+	not has_securitycontext_pod
+	profile := input.review.object.metadata.annotations[pod_annotation_key]
+	location := sprintf("annotation %v", [pod_annotation_key])
+}
+
+# Container profile as defined in container annotation
+get_profile(container) = {"profile": profile, "location": location} {
+	not has_securitycontext_container(container)
+	container_annotation := get_container_annotation_key(container.name)
+	has_annotation(container_annotation)
+	profile := input.review.object.metadata.annotations[container_annotation]
+	location := sprintf("annotation %v", [container_annotation])
+}
+
+# Container profile as defined in pods securityContext
+get_profile(container) = {"profile": profile, "location": location} {
+	not has_securitycontext_container(container)
+	not has_annotation(get_container_annotation_key(container.name))
+	profile := input.review.object.spec.securityContext.seccompProfile.type
+	location := "pod securityContext"
+}
+
+# Container profile as defined in containers securityContext
+get_profile(container) = {"profile": profile, "location": location} {
+	has_securitycontext_container(container)
+	profile := container.securityContext.seccompProfile.type
+	location := "container securityContext"
```

Comment:
`Integration test on Gatekeeper release-3.6` is failing here when creating [library/pod-security-policy/seccomp/samples/psp-seccomp/example_allowed.yaml](https://github.com/open-policy-agent/gatekeeper-library/blob/master/library/pod-security-policy/seccomp/samples/psp-seccomp/example_allowed.yaml).  This should not be happening because the yaml does not set `securityContext.seccompProfile`.  How do we debug the integration test to see what the object being presented to k8s looks like?

Reply:
I have investigated this issue, looks like in a live cluster, the `kube-apiserver` will translate the annotation to securityContext. So the example file:
```
apiVersion: v1
kind: Pod
metadata:
  name: nginx-seccomp-allowed
  annotations:
    container.seccomp.security.alpha.kubernetes.io/nginx: runtime/default
  labels:
    app: nginx-seccomp
spec:
  containers:
  - name: nginx
    image: nginx
```
results in the input with the following object (trimmed):
```
"object": {
      "apiVersion": "v1",
      "kind": "Pod",
      "metadata": {
        "annotations": {
          "container.seccomp.security.alpha.kubernetes.io/nginx": "runtime/default"
        },
        "creationTimestamp": "2022-02-04T15:12:26Z",
        "labels": {
          "app": "nginx-seccomp"
        },
        "name": "nginx-seccomp-allowed",
        "namespace": "policy-review-exporting",
        "uid": "d40d3dc3-fe9d-433f-beae-50d2b275c4f2"
      },
      "spec": {
        "containers": [
          {
            "image": "nginx",
            "imagePullPolicy": "Always",
            "name": "nginx",
            "resources": {},
            "securityContext": {
              "seccompProfile": {
                "type": "RuntimeDefault"
              }
            },
            "terminationMessagePath": "/dev/termination-log",
            "terminationMessagePolicy": "File",
            "volumeMounts": [
              {
                "mountPath": "/var/run/secrets/kubernetes.io/serviceaccount",
                "name": "kube-api-access-66kfc",
                "readOnly": true
              }
            ]
          }
        ],
        "dnsPolicy": "ClusterFirst",
        "enableServiceLinks": true,
        "preemptionPolicy": "PreemptLowerPriority",
        "priority": 0,
        "restartPolicy": "Always",
        "schedulerName": "default-scheduler",
        "securityContext": {},
        "serviceAccount": "default",
        "serviceAccountName": "default",
        "terminationGracePeriodSeconds": 30,
        "tolerations": [
          {
            "effect": "NoExecute",
            "key": "node.kubernetes.io/not-ready",
            "operator": "Exists",
            "tolerationSeconds": 300
          },
          {
            "effect": "NoExecute",
            "key": "node.kubernetes.io/unreachable",
            "operator": "Exists",
            "tolerationSeconds": 300
          }
        ],
        "volumes": [
          {
            "name": "kube-api-access-66kfc",
            "projected": {
              "defaultMode": 420,
              "sources": [
                {
                  "serviceAccountToken": {
                    "expirationSeconds": 3607,
                    "path": "token"
                  }
                },
                {
                  "configMap": {
                    "items": [
                      {
                        "key": "ca.crt",
                        "path": "ca.crt"
                      }
                    ],
                    "name": "kube-root-ca.crt"
                  }
                },
                {
                  "downwardAPI": {
                    "items": [
                      {
                        "fieldRef": {
                          "apiVersion": "v1",
                          "fieldPath": "metadata.namespace"
                        },
                        "path": "namespace"
                      }
                    ]
                  }
                }
              ]
            }
          }
        ]
      },
      "status": {
        "phase": "Pending",
        "qosClass": "BestEffort"
      }
    },
```

Notice that the container has the appropriate securityContext in the admission review.

I think for backwards compatibility I may have to prefer the annotation over the securityContext, what do you think @maxsmythe and @swarren83 ?

Reply:
Yes, I think the annotation needs to be evaluated first and then seccompProfile.

Reply:
I have run the integration tests manually successfully with this modification, I will push it to the branch.
```
$ make test-integration
bats -t test/bats/test.bats
1..8
ok 1 gatekeeper-controller-manager is running
ok 2 gatekeeper-audit is running
ok 3 namespace label webhook is serving
ok 4 constrainttemplates crd is established
ok 5 waiting for validating webhook
ok 6 applying sync config
ok 7 waiting for namespaces to be synced using metrics endpoint
ok 8 testing constraint templates
```